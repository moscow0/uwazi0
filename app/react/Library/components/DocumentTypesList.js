import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Immutable, { is } from 'immutable';
import ShowIf from 'app/App/ShowIf';
import { t, Translate } from 'app/I18N';
import { Icon } from 'UI';

import { filterDocumentTypes } from 'app/Library/actions/filterActions';

const getItemsToShow = (fromFilters, templates, settings) => {
  let items = fromFilters ? settings.collection.toJS().filters : [];
  if (!items?.length) {
    items = templates.toJS().map(tpl => ({
      id: tpl._id,
      name: tpl.name,
    }));
  }
  return items;
};

export class DocumentTypesList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ui: {},
    };
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      !is(this.props.selectedTemplates, nextProps.selectedTemplates) ||
      !is(this.props.settings, nextProps.settings) ||
      !is(this.props.aggregations, nextProps.aggregations) ||
      this.stateChanged(nextState)
    );
  }

  changeAll(item, e) {
    const { selectedTemplates: selectedItems } = this.props;
    if (e.target.checked) {
      item.items.forEach(_item => {
        if (!this.checked(_item)) {
          selectedItems.push(_item.id);
        }
      });
    }

    if (!e.target.checked) {
      item.items.forEach(_item => {
        if (this.checked(_item)) {
          const index = selectedItems.indexOf(_item.id);
          selectedItems.splice(index, 1);
        }
      });
    }

    this.props.filterDocumentTypes(selectedItems);
  }

  change(item) {
    const { selectedTemplates: selectedItems } = this.props;

    if (selectedItems.includes(item.id)) {
      const index = selectedItems.indexOf(item.id);
      selectedItems.splice(index, 1);
    } else {
      selectedItems.push(item.id);
    }

    this.props.filterDocumentTypes(selectedItems);
  }

  toggleOptions(item, e) {
    e.preventDefault();
    if (!this.checked(item) && item.items.find(itm => this.checked(itm))) {
      return;
    }
    const { ui } = this.state;
    ui[item.id] = !ui[item.id];
    this.setState({ ui });
  }

  aggregations(item) {
    const aggregations = this.aggs;
    const buckets =
      aggregations.all && aggregations.all._types ? aggregations.all._types.buckets : [];
    const found = buckets.find(agg => agg.key === item.id);
    if (found) {
      return found.filtered.doc_count;
    }

    if (item.items) {
      return item.items.reduce((result, _item) => result + this.aggregations(_item), 0);
    }

    return 0;
  }

  showSubOptions(parent) {
    const toggled = this.state.ui[parent.id];
    return !!(toggled || !!(!this.checked(parent) && parent.items.find(itm => this.checked(itm))));
  }

  checked(item) {
    if (item.items) {
      return item.items.reduce(
        (result, _item) => result && this.checked(_item),
        item.items.length > 0
      );
    }

    return this.props.selectedTemplates.includes(item.id);
  }

  stateChanged(nextState) {
    return (
      Object.keys(nextState.ui).length === Object.keys(this.state.ui).length ||
      Object.keys(nextState.ui).reduce(
        (result, key) => result || nextState.ui[key] === this.state.ui[key],
        false
      )
    );
  }

  renderSingleType(item, index) {
    const context = item.id === 'missing' ? 'System' : item.id;
    return (
      <li className="multiselectItem" key={index} title={item.name}>
        <input
          type="checkbox"
          className="multiselectItem-input"
          value={item.id}
          id={item.id}
          onChange={this.change.bind(this, item)}
          checked={this.checked(item)}
        />
        <label className="multiselectItem-label" htmlFor={item.id}>
          <span className="multiselectItem-icon">
            <Icon icon={['far', 'square']} className="checkbox-empty" />
            <Icon icon="check" className="checkbox-checked" />
          </span>
          <span className="multiselectItem-name">
            <Translate context={context}>{item.name}</Translate>
          </span>
        </label>
        <span className="multiselectItem-results">{this.aggregations(item)}</span>
      </li>
    );
  }

  renderGroupType(item, index) {
    return (
      <li key={index}>
        <div className="multiselectItem">
          <input
            type="checkbox"
            className="form-control multiselectItem-input"
            id={item.id}
            onChange={this.changeAll.bind(this, item)}
            checked={this.checked(item)}
          />
          <label htmlFor={item.id} className="multiselectItem-label">
            <span className="multiselectItem-icon">
              <Icon icon={['far', 'square']} className="checkbox-empty" />
              <Icon icon="check" className="checkbox-checked" />
            </span>
            <span className="multiselectItem-name">
              <b>{t('Filters', item.name)}</b>
            </span>
          </label>
          <span className="multiselectItem-results">
            <span>{this.aggregations(item)}</span>
            <span className="multiselectItem-action" onClick={this.toggleOptions.bind(this, item)}>
              <Icon icon={this.state.ui[item.id] ? 'caret-up' : 'caret-down'} />
            </span>
          </span>
        </div>
        <ShowIf if={this.showSubOptions(item)}>
          <ul className="multiselectChild is-active">
            {item.items.map((_item, i) => this.renderSingleType(_item, i))}
          </ul>
        </ShowIf>
      </li>
    );
  }

  render() {
    const { fromFilters, templates, settings } = this.props;
    const items = getItemsToShow(fromFilters, templates, settings);
    this.aggs = this.props.aggregations.toJS();
    return (
      <ul className="multiselect is-active">
        {items.map((item, index) => {
          if (item.items) {
            return this.renderGroupType(item, index);
          }
          return this.renderSingleType(item, index);
        })}
      </ul>
    );
  }
}

DocumentTypesList.defaultProps = {
  fromFilters: true,
  templates: Immutable.fromJS([]),
  settings: {},
  aggregations: Immutable.fromJS({}),
  selectedTemplates: [],
  filterDocumentTypes: {},
};

DocumentTypesList.propTypes = {
  selectedTemplates: PropTypes.instanceOf(Array),
  settings: PropTypes.instanceOf(Object),
  templates: PropTypes.instanceOf(Immutable.List),
  filterDocumentTypes: PropTypes.func,
  aggregations: PropTypes.instanceOf(Immutable.Map),
  fromFilters: PropTypes.bool,
};

export function mapStateToProps(state) {
  return {
    settings: state.settings,
    templates: state.templates,
    aggregations: state.library.aggregations,
  };
}

//parent ts component requires ownProps to inferring types
function mapDispatchToProps(dispatch, _ownProps) {
  return bindActionCreators({ filterDocumentTypes }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(DocumentTypesList);
