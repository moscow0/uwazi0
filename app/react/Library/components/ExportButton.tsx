import React, { Component } from 'react';
import { Icon } from 'app/UI';
import { t, Translate } from 'app/I18N';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { wrapDispatch } from 'app/Multireducer';
import { exportDocuments } from 'app/Library/actions/exportActions';
import { User } from 'api/users/usersModel';
import Modal from 'app/Layout/Modal';
import { LocalForm } from 'react-redux-form';
import { CaptchaValue } from 'shared/types/Captcha';
import { Captcha, FormGroup } from 'app/ReactReduxForms';
import { IImmutable } from 'shared/types/Immutable';
import { ExportStore } from '../reducers/ExportStoreType';

type ExportButtonProps = {
  processing: boolean;
  storeKey: string;
  user: IImmutable<User>;
  className?: string;
  exportDocuments: (keyStore: string, captcha?: CaptchaValue) => any;
};

class ExportButton extends Component<ExportButtonProps, { modal: boolean }> {
  static defaultProps: { className: string };

  constructor(props: ExportButtonProps) {
    super(props);
    this.state = {
      modal: false,
    };
    this.export = this.export.bind(this);
    this.showModal = this.showModal.bind(this);
    this.exportWithCaptcha = this.exportWithCaptcha.bind(this);
  }

  export() {
    if (!this.props.processing) {
      this.props.exportDocuments(this.props.storeKey);
    }
  }

  exportWithCaptcha(values: { captcha: CaptchaValue }) {
    if (!this.props.processing) {
      this.props.exportDocuments(this.props.storeKey, values.captcha);
    }
    this.setState({ modal: false });
  }

  showModal() {
    this.setState({ modal: true });
  }

  render() {
    let btnClassName = 'btn btn-default btn-export';
    btnClassName += this.props.processing ? ' btn-disabled' : '';
    btnClassName += ` ${this.props.className}`;
    return (
      <>
        <button
          type="button"
          onClick={this.props.user.get('_id') ? this.export : this.showModal}
          className={btnClassName}
        >
          {!this.props.processing ? (
            <Icon icon="export-csv" transform="up-0.1" />
          ) : (
            <Icon icon="spinner" spin />
          )}
          <span className="btn-label">{t('System', 'Export CSV')}</span>
        </button>
        {this.state.modal && (
          <Modal isOpen type="export">
            <LocalForm
              onSubmit={this.exportWithCaptcha}
              validators={{ captcha: { required: (val: any) => val && val.text.length } }}
            >
              <FormGroup key="captcha" model=".captcha">
                <h3>
                  <Translate>Exporting entities to CSV</Translate>
                </h3>
                <p>
                  <Translate>
                    Please type in letters and numbers from the image to start the export.
                  </Translate>
                </p>
                <Captcha remote={false} model=".captcha" />
              </FormGroup>
              <div className="buttons">
                <input
                  type="button"
                  className="btn"
                  onClick={() => {
                    this.setState({ modal: false });
                  }}
                  value="Cancel"
                />
                <input type="submit" className="btn btn-success" value="Export" />
              </div>
            </LocalForm>
          </Modal>
        )}
      </>
    );
  }
}

ExportButton.defaultProps = {
  className: '',
};

function mapDispatchToProps(dispatch: Dispatch<any>, props: Pick<ExportButtonProps, 'storeKey'>) {
  return bindActionCreators({ exportDocuments }, wrapDispatch(dispatch, props.storeKey));
}

function mapStateToProps(state: ExportStore) {
  return {
    processing: state.exportSearchResults.exportSearchResultsProcessing,
    user: state.user,
  };
}

export type { ExportButtonProps };

export default connect(mapStateToProps, mapDispatchToProps)(ExportButton);
