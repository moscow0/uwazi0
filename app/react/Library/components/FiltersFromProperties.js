import { connect } from 'react-redux';
import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React from 'react';

import { t } from 'app/I18N';
import FormGroup from 'app/DocumentForm/components/FormGroup';
import { getAggregationSuggestions } from 'app/Library/actions/libraryActions';
import { selectTemplates } from 'app/utils/coreSelectors';
import DateFilter from './DateFilter';
import NestedFilter from './NestedFilter';
import NumberRangeFilter from './NumberRangeFilter';
import SelectFilter from './SelectFilter';
import TextFilter from './TextFilter';

const prepareOptions = property => {
  const filteredProperty = {
    ...property,
    options: property.options.filter(option => option.id !== 'any'),
  };
  return filteredProperty.options.map(option => {
    const finalTranslatedOption = {
      ...option,
      label: t(filteredProperty.content, option.label, undefined, false),
    };

    if (option.options) {
      const translatedSubOptions = option.options.map(subOption => ({
        ...subOption,
        label: t(filteredProperty.content, subOption.label, undefined, false),
      }));
      finalTranslatedOption.options = translatedSubOptions;
    }
    return finalTranslatedOption;
  });
};

export const FiltersFromProperties = ({
  onChange,
  properties,
  translationContext,
  modelPrefix = '',
  storeKey,
  templates,
  ...props
}) => (
  <>
    {properties.map(property => {
      const { type } = property.inherit?.property ? property.inherit : property;

      const commonProps = {
        model: `.filters${modelPrefix}.${property.name}`,
        label: t(translationContext, property.label),
        onChange,
      };

      const propertyOptions = property.options ? prepareOptions(property) : [];

      let filter = <TextFilter {...commonProps} />;

      if (type === 'numeric') {
        filter = <NumberRangeFilter {...commonProps} />;
      }

      if (['select', 'multiselect', 'relationship'].includes(type)) {
        filter = (
          <SelectFilter
            {...commonProps}
            lookup={getAggregationSuggestions.bind(null, storeKey, property.name)}
            options={propertyOptions}
            prefix={property.name}
            showBoolSwitch={property.type === 'multiselect' || property.type === 'relationship'}
            sort={property.type === 'relationship'}
            totalPossibleOptions={property.totalPossibleOptions}
            allowSelectGroup
          />
        );
      }

      if (type === 'nested') {
        filter = (
          <NestedFilter {...commonProps} property={property} aggregations={props.aggregations} />
        );
      }

      if (
        type === 'date' ||
        type === 'multidate' ||
        type === 'multidaterange' ||
        type === 'daterange'
      ) {
        filter = <DateFilter {...commonProps} format={props.dateFormat} />;
      }

      return <FormGroup key={property.name}>{filter}</FormGroup>;
    })}
  </>
);

FiltersFromProperties.defaultProps = {
  onChange: () => {},
  dateFormat: '',
  modelPrefix: '',
  translationContext: '',
};

FiltersFromProperties.propTypes = {
  templates: PropTypes.instanceOf(Array).isRequired,
  onChange: PropTypes.func,
  dateFormat: PropTypes.string,
  modelPrefix: PropTypes.string,
  translationContext: PropTypes.string,
  storeKey: PropTypes.string.isRequired,
  aggregations: PropTypes.instanceOf(Immutable.Map).isRequired,
  properties: PropTypes.array.isRequired,
};

export function mapStateToProps(state, props) {
  return {
    dateFormat: state.settings.collection.get('dateFormat'),
    aggregations: state[props.storeKey].aggregations,
    storeKey: props.storeKey,
    templates: selectTemplates(state),
  };
}

export default connect(mapStateToProps)(FiltersFromProperties);
