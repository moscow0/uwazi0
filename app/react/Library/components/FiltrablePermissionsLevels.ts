export const filtrableLevels = ['read', 'write'] as const;
export type FiltrableLevel = typeof filtrableLevels[number];
