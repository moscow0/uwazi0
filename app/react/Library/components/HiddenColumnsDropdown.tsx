import React, { RefObject, useCallback, useRef, useState } from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { List } from 'immutable';
import { DropdownList } from 'app/Forms';
import { TableViewColumn, IStore } from 'app/istore';
import { wrapDispatch } from 'app/Multireducer';
import {
  ColumnItem,
  ValueItem,
  SelectableColumn,
} from 'app/Library/components/HiddenColumnsDropdownItem';
import {
  setTableViewColumnHidden,
  setTableViewAllColumnsHidden,
} from 'app/Library/actions/libraryActions';
import { IImmutable } from 'shared/types/Immutable';
import { useOnClickOutsideElement } from 'app/utils/useOnClickOutsideElementHook';

interface HideColumnsComponentProps {
  columns: List<IImmutable<TableViewColumn>>;
  setTableViewColumnHidden: (name: string, hidden: boolean) => void;
  setTableViewAllColumnsHidden: (hidden: boolean) => void;
}
const mapStateToProps = (state: IStore) => ({
  columns: state.library.ui.get('tableViewColumns'),
});
const mapDispatchToProps = (dispatch: Dispatch<IStore>) =>
  bindActionCreators(
    { setTableViewColumnHidden, setTableViewAllColumnsHidden },
    wrapDispatch(dispatch, 'library')
  );
const connector = connect(mapStateToProps, mapDispatchToProps);

const processColumns = (
  columnsMap: List<{
    toJS(): TableViewColumn;
    get<Field extends keyof TableViewColumn>(_field: Field): IImmutable<TableViewColumn[Field]>;
    filter(fn: (listElement: any) => boolean | undefined): List<any>;
  }>
) => {
  const columns = columnsMap.toJS().slice(1);
  const hiddenColumns = columns.filter((c: TableViewColumn) => c.hidden);
  const shownColumns = columns.filter((c: TableViewColumn) => !c.hidden);
  const selectAllColumn: Partial<SelectableColumn> = {
    label: 'Show all',
    selectAll: true,
    indeterminate: hiddenColumns.length !== 0 && shownColumns.length !== 0,
    hidden: shownColumns.length === 0,
    type: 'text',
  };
  const sortedColumns = [selectAllColumn].concat(
    shownColumns.concat(hiddenColumns).map((c: TableViewColumn) => ({ ...c, selectAll: false }))
  );
  return { sortedColumns, hiddenColumns };
};

export const HideColumnsComponent = ({
  setTableViewAllColumnsHidden: setAllColumnsHidden,
  setTableViewColumnHidden: setColumnHidden,
  columns: columnsMap,
}: HideColumnsComponentProps) => {
  const [open, setOpen] = useState(false);
  const [clickedOutside, setClickedOutside] = useState(false);

  const { sortedColumns, hiddenColumns } = processColumns(columnsMap);
  const dropdownContainerRef = useRef(null);
  const dropdownRef: RefObject<React.Component & React.ReactElement> = useRef(null);

  const onClickOutside = useCallback(event => {
    if (
      event.target.className !== 'columns-hint' &&
      (!event.target.parentElement || event.target.parentElement.className !== 'columns-hint')
    ) {
      setClickedOutside(true);
      dropdownRef.current?.props.onToggle();
    }
  }, []);

  const handleClick = () => {
    if (open) {
      setOpen(false);
    }
  };

  useOnClickOutsideElement<HTMLLIElement>(dropdownContainerRef, onClickOutside);

  const onSelect = (item: SelectableColumn) =>
    item.selectAll
      ? setAllColumnsHidden(item.indeterminate ? false : !item.hidden)
      : setColumnHidden(item.name, !item.hidden);

  return (
    <div className="hidden-columns-dropdown" ref={dropdownContainerRef}>
      {/*
        // @ts-ignore */}
      <DropdownList
        ref={dropdownRef}
        open={open}
        data={sortedColumns}
        filter={(item: SelectableColumn, searchTerm: string) =>
          item.label.toLowerCase().includes(searchTerm.toLowerCase())
        }
        itemComponent={ColumnItem}
        valueComponent={ValueItem(hiddenColumns, open, handleClick)}
        onSelect={(selected: SelectableColumn) => {
          onSelect(selected);
        }}
        onToggle={() => {
          if (clickedOutside) {
            setOpen(false);
            setClickedOutside(false);
            dropdownRef.current?.forceUpdate();
            return;
          }
          if (!open) {
            setOpen(true);
          }
        }}
      />
    </div>
  );
};
export const HiddenColumnsDropdown = connector(HideColumnsComponent);
