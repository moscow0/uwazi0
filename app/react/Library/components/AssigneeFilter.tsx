import React, { useCallback, useMemo } from 'react';
import FormGroup from 'app/DocumentForm/components/FormGroup';
import { Aggregations } from 'shared/types/aggregations';
import { NeedAuthorization } from 'app/Auth';
import { Translate } from 'app/I18N';
import { MultiSelect } from 'app/Forms';
import { Control } from 'react-redux-form';
import { Switcher } from 'app/ReactReduxForms';
import { FiltrableLevel } from './FiltrablePermissionsLevels';

interface PermissionsValue {
  level: FiltrableLevel;
  refId: string;
}

interface AssigneeFilterProps {
  onChange: (values: PermissionsValue[]) => void;
  aggregations: Aggregations;
}

const generateOptions = (aggregations: Aggregations, level: FiltrableLevel) =>
  aggregations?.all?.[`_permissions.${level}`]?.buckets
    .filter(aggregation => aggregation.key !== 'any')
    .map(aggregation => ({
      label: aggregation.label!,
      title: aggregation.label!,
      value: aggregation.key,
      results: aggregation.filtered.doc_count,
      ...(aggregation.icon ? { icon: { type: 'Icons', _id: aggregation.icon } } : {}),
    })) || [];

export const AssigneeFilterSelectUncontrolled = ({
  onChange,
  aggregations,
  value = [],
}: AssigneeFilterProps & { value: PermissionsValue[] }) => {
  const readOptions = useMemo(() => generateOptions(aggregations, 'read'), [aggregations]);
  const writeOptions = useMemo(() => generateOptions(aggregations, 'write'), [aggregations]);

  const getChangeHandler = (level: FiltrableLevel) => (values: string[]) => {
    const newSelected = value
      .filter((v: PermissionsValue) => v.level !== level)
      .concat(values.map(v => ({ refId: v, level })));

    onChange(newSelected);
  };

  const onReadChangeHandler = useCallback(getChangeHandler('read'), [value, onChange]);
  const onWriteChangeHandler = useCallback(getChangeHandler('write'), [value, onChange]);

  const readValues = value.filter(v => v.level === 'read').map(v => v.refId);
  const writeValues = value.filter(v => v.level === 'write').map(v => v.refId);

  return (
    <>
      <li>
        <Translate>Shared as viewer</Translate>
      </li>
      <li className="wide">
        <MultiSelect
          prefix="permissions.read"
          onChange={onReadChangeHandler}
          options={readOptions}
          value={readValues}
        />
      </li>
      <li className="spaced">
        <Translate>Shared as editor</Translate>
      </li>
      <li className="wide">
        <MultiSelect
          prefix="permissions.write"
          onChange={onWriteChangeHandler}
          options={writeOptions}
          value={writeValues}
        />
      </li>
    </>
  );
};

// eslint-disable-next-line react/no-multi-comp
const AssigneeFilterSelect = ({
  model,
  prefix,
  onChange,
  aggregations,
}: AssigneeFilterProps & { model: string; prefix: string }) => (
  // eslint-disable-next-line react/jsx-pascal-case
  <Control.select
    component={AssigneeFilterSelectUncontrolled}
    model={model}
    prefix={prefix}
    onChange={onChange as any}
    // @ts-ignore
    aggregations={aggregations}
  />
);

// eslint-disable-next-line react/no-multi-comp
export const AssigneeFilter = ({ onChange, aggregations }: AssigneeFilterProps) => (
  <NeedAuthorization>
    <FormGroup key="permissions" className="admin-filter">
      <ul className="search__filter is-active">
        <li>
          <Translate>Permissions</Translate>
          <Switcher
            model={".customFilters['permissions'].and"}
            prefix="permissions"
            onChange={onChange}
          />
        </li>
        <AssigneeFilterSelect
          model=".customFilters['permissions'].values"
          prefix="permissions"
          onChange={onChange}
          aggregations={aggregations}
        />
      </ul>
    </FormGroup>
  </NeedAuthorization>
);
