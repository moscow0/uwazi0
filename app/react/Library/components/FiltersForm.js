import { Form } from 'react-redux-form';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Immutable, { is, fromJS } from 'immutable';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Icon } from 'UI';

import debounce from 'app/utils/debounce';
import libraryHelper from 'app/Library/helpers/libraryFilters';
import { searchDocuments } from 'app/Library/actions/libraryActions';
import { Translate } from 'app/I18N';
import { wrapDispatch } from 'app/Multireducer';
import { FilterTocGeneration } from 'app/ToggledFeatures/tocGeneration';
import { TemplatesFilter } from 'app/Library/components/TemplatesFilter';
import { AssigneeFilter } from 'app/Library/components/AssigneeFilter';
import { PermissionsFilter } from './PermissionsFilter';
import { PublishedFilters } from './PublishedFilters';

import Filters from './FiltersFromProperties';

export class FiltersForm extends Component {
  constructor(props) {
    super(props);
    this.search = debounce(search => {
      this.props.searchDocuments({ search }, this.props.storeKey);
    }, 300);

    this.submit = this.submit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.activateAutoSearch = () => {
      this.autoSearch = true;
    };

    this.state = { documentTypeFromFilters: true };
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      !is(this.props.fields, nextProps.fields) ||
      !is(this.props.aggregations, nextProps.aggregations) ||
      !is(this.props.documentTypes, nextProps.documentTypes) ||
      !is(this.state.documentTypeFromFilters, nextState.documentTypeFromFilters)
    );
  }

  onChange(search) {
    if (this.autoSearch) {
      this.autoSearch = false;
      this.search(search, this.props.storeKey);
    }
  }

  submit(search) {
    this.props.searchDocuments({ search }, this.props.storeKey);
  }

  render() {
    const { templates, documentTypes } = this.props;
    const aggregations = this.props.aggregations.toJS();
    const translationContext =
      documentTypes.get(0) || (templates.get(0) || fromJS({})).get('_id') || 'System';
    const allFields = this.props.fields.toJS();
    const showNoValueOnFilters = allFields.size;
    const fields = libraryHelper
      .parseWithAggregations(allFields.slice(0), aggregations, showNoValueOnFilters)
      .filter(field => !field.options || field.options.length);
    const model = `${this.props.storeKey}.search`;

    return (
      <div className="filters-box">
        <Form model={model} id="filtersForm" onSubmit={this.submit} onChange={this.onChange}>
          <PublishedFilters onChange={this.activateAutoSearch} aggregations={aggregations} />
          <PermissionsFilter onChange={this.activateAutoSearch} aggregations={aggregations} />
          <TemplatesFilter />
          <Filters
            onChange={this.activateAutoSearch}
            properties={fields}
            translationContext={translationContext}
            storeKey={this.props.storeKey}
          />

          <FilterTocGeneration onChange={this.activateAutoSearch} aggregations={aggregations} />
          <AssigneeFilter onChange={this.activateAutoSearch} aggregations={aggregations} />
        </Form>

        {(() => {
          const activeTypes = templates.filter(template =>
            documentTypes.includes(template.get('_id'))
          );
          if (activeTypes.size > 0 && fields.length === 0) {
            return (
              <div className="blank-state">
                <Icon icon="times" />
                <h4>
                  <Translate>No common filters</Translate>
                </h4>
                <p>
                  <Translate>
                    The combination of entity types doesn&#39;t have any filters in common.
                  </Translate>
                </p>
                <a
                  href="https://github.com/huridocs/uwazi/wiki/Filter"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <Translate>Learn more</Translate>
                </a>
              </div>
            );
          }

          return null;
        })()}
      </div>
    );
  }
}

FiltersForm.propTypes = {
  templates: PropTypes.instanceOf(Immutable.List).isRequired,
  aggregations: PropTypes.instanceOf(Immutable.Map).isRequired,
  fields: PropTypes.instanceOf(Immutable.List).isRequired,
  searchDocuments: PropTypes.func.isRequired,
  search: PropTypes.object,
  documentTypes: PropTypes.instanceOf(Immutable.List).isRequired,
  storeKey: PropTypes.string.isRequired,
};

export function mapStateToProps(state, props) {
  return {
    fields: state[props.storeKey].filters.get('properties'),
    aggregations: state[props.storeKey].aggregations,
    templates: state.templates,
    documentTypes: state[props.storeKey].filters.get('documentTypes'),
  };
}

function mapDispatchToProps(dispatch, props) {
  return bindActionCreators({ searchDocuments }, wrapDispatch(dispatch, props.storeKey));
}

export default connect(mapStateToProps, mapDispatchToProps)(FiltersForm);
