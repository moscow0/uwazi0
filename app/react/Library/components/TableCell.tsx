import React from 'react';
import { MetadataObjectSchema, PropertySchema } from 'shared/types/commonTypes';
import { showByType } from 'app/Metadata/components/Metadata';

interface TableCellProps {
  content?: FormattedMetadataValue;
  zoomLevel: number;
}

type FormattedMetadataValue = Omit<PropertySchema, 'type'> & {
  inheritedName?: string;
  value?:
    | string
    | MetadataObjectSchema
    | MetadataObjectSchema[]
    | [{ parent: string; value: MetadataObjectSchema[] }];
  type: 'inherit' | PropertySchema['type'];
  inheritedType?: PropertySchema['type'];
  onlyForCards?: boolean;
};

const formatProperty = (prop: FormattedMetadataValue | undefined) => {
  if (!prop?.value) {
    return undefined;
  }

  const inheritedMedia =
    prop.type === 'inherit' && (prop.inheritedType === 'image' || prop.inheritedType === 'media');
  const typeMedia = prop.type === 'image' || prop.type === 'media';

  if (inheritedMedia || typeMedia) {
    return undefined;
  }
  return showByType(prop, true);
};

export const TableCellComponent = (props: TableCellProps) => {
  const cellValue = formatProperty(props.content);
  return (
    <div className={`table-view-cell table-view-row-zoom-${props.zoomLevel}`}>{cellValue}</div>
  );
};

export const TableCell = React.memo(TableCellComponent);

export type { TableCellProps, FormattedMetadataValue };
