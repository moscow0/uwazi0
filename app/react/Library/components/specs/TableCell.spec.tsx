/**
 * @jest-environment jsdom
 */

import { TableCell, TableCellProps } from 'app/Library/components/TableCell';
import { shallow } from 'enzyme';
import React from 'react';
import MarkdownViewer from 'app/Markdown';
import ValueList from 'app/Metadata/components/ValueList';

describe('TableCell', () => {
  let component: any;
  const props: TableCellProps = {
    content: { label: 'Title', type: 'text', name: 'title', value: 'Entity 1' },
    zoomLevel: 1,
  };

  function renderContent() {
    component = shallow(<TableCell {...props} />);
    const cells = component.find('.table-view-cell');
    return cells.at(0);
  }

  it.each`
    label       | type             | name         | value
    ${'Title'}  | ${'text'}        | ${'title'}   | ${'Title 1'}
    ${'Date'}   | ${'date'}        | ${'date'}    | ${'May 20, 2019'}
    ${'AutoId'} | ${'generatedid'} | ${'auto_id'} | ${'XYZ-1234'}
  `(
    'should render a plain content with the passed value for $type',
    ({ label, type, name, value }) => {
      props.content = { label, type, name, value };
      const cellContent = renderContent();
      expect(cellContent.props().children).toBe(value);
    }
  );

  it('should render a date with as value passed', () => {
    const formattedPropertyDate = 'May 20, 2019';
    props.content = { label: 'Date', type: 'date', name: 'date', value: formattedPropertyDate };
    const cellContent = renderContent();
    expect(cellContent.props().children).toBe(formattedPropertyDate);
  });

  it('should render a multi select as a list of values', () => {
    props.content = {
      label: 'Languages',
      type: 'multiselect',
      name: 'languages',
      value: [{ value: 'EspaÃ±ol' }, { value: 'English' }],
    };
    const cellContent = renderContent();
    expect(cellContent.props().children).toMatchSnapshot();
  });

  it('should render a multi select of nested values as a list of values', () => {
    props.content = {
      label: 'People',
      type: 'multiselect',
      name: 'people',
      value: [{ parent: 'male', value: [{ value: 'A guy' }] }],
    };
    const cellContent = renderContent();
    expect(cellContent.props().children).toMatchSnapshot();
  });

  it('should render an external link', () => {
    props.content = {
      label: 'Google',
      type: 'link',
      name: 'google',
      value: {
        label: 'google',
        url: 'www.google.com',
        value: null,
      },
    };
    const cellContent = renderContent();
    expect(cellContent.find('a').props().href).toBe('www.google.com');
  });

  it('should render a relationship as a list of links', () => {
    props.content = {
      label: 'Relationship',
      type: 'relationship',
      name: 'relationship',
      value: [
        {
          url: '/entity/Entity1',
          value: 'Entity1',
        },
        {
          url: '/entity/Entity2',
          value: 'Entity2',
        },
      ],
    };
    const cellContent = renderContent();
    const firstLink =
      cellContent.props().children.props.property.value[0].value.props.propValue.url;
    const secondLink =
      cellContent.props().children.props.property.value[1].value.props.propValue.url;
    expect(firstLink).toEqual('/entity/Entity1');
    expect(secondLink).toEqual('/entity/Entity2');
  });

  it('should render inherited properties', () => {
    props.content = {
      inheritedName: 'sexo',
      inheritedType: 'select',
      label: 'Relationship',
      type: 'inherit',
      name: 'relationship',
      value: [{ label: 'Sexo', name: 'sexo', value: 'Mujer' }],
    };
    const cellContent = renderContent();
    expect(cellContent.find(ValueList).props().property.value).toEqual(props.content.value);
  });

  it('should render a geolocation as a compact view', () => {
    props.content = {
      label: 'Geolocation',
      type: 'geolocation',
      name: 'geolocation',
      value: [{ lon: 2, lat: 46, value: null }],
      onlyForCards: true,
    };

    const cellContent = renderContent();
    const geolocationProps = cellContent.props().children.props;

    expect(geolocationProps.points).toEqual([{ lon: 2, lat: 46, value: null }]);
    expect(geolocationProps.onlyForCards).toBe(true);
  });

  it('should render a rich text as a markdown', () => {
    props.content = {
      label: 'Rich Text',
      type: 'markdown',
      name: 'rich_text',
      value: '**bold**',
    };
    const cellContent = renderContent();
    expect(cellContent.find(MarkdownViewer).props().markdown).toBe('**bold**');
  });

  it('should not render if value is empty', () => {
    props.content = {
      label: 'Empty Select',
      type: 'select',
      name: 'empty_select',
    };
    const cellContent = renderContent();
    expect(cellContent.props().children).toBe(undefined);
  });

  it('should not render entity has not value for column', () => {
    delete props.content;
    const cellContent = renderContent();
    expect(cellContent.props().children).toBe(undefined);
  });

  it('should not render if the type is not supported', () => {
    props.content = {
      label: 'Non supported Image',
      type: 'image',
      name: 'non_supported_image',
      value: [{ value: 'url' }],
    };
    const cellContent = renderContent();
    expect(cellContent.props().children).toEqual(undefined);
  });
});
