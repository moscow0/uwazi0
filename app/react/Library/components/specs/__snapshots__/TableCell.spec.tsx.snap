// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableCell should render a multi select as a list of values 1`] = `
<ValueList
  compact={true}
  property={
    Object {
      "label": "Languages",
      "name": "languages",
      "type": "multiselect",
      "value": Array [
        Object {
          "value": "EspaÃ±ol",
        },
        Object {
          "value": "English",
        },
      ],
    }
  }
/>
`;

exports[`TableCell should render a multi select of nested values as a list of values 1`] = `
<ValueList
  compact={true}
  property={
    Object {
      "label": "People",
      "name": "people",
      "type": "multiselect",
      "value": Array [
        Object {
          "value": "male: A guy",
        },
      ],
    }
  }
/>
`;
